File: code/fsharp-visualstudio/Solutions/03-RevisedFarmModel.fs
module Solutions.RevisedFarmModel

open Flips
open Flips.Types

let solve () =

  // Define the set of animals you are working with
  let animals = ["cow"; "pig"; "chicken"]

  // Dictionary of Revenue indexed by Animal Type
  let revenue = ["cow", 100.0; "pig", 50.0; "chicken", 8.0] |> Map.ofList

  // Dictionary of Pasture required per animal indexed by Animal Type
  let pasture = ["cow",1.0; "pig",0.5; "chicken",0.0] |> Map.ofList

  // Dictionary of Feed required per animal indexed by Animal Type
  let feed = ["cow",0.0; "pig",1.0; "chicken",0.1] |> Map.ofList

  // Dictionary of the Labor required per animal indexed by Animal Type
  let labor = ["cow",9.5; "pig",8.0; "chicken",0.0] |> Map.ofList

  // Constants which determine the limits on our resources
  let totalPasture = 1000.0
  let totalFeed = 100.0
  let totalLabor = 7200.0


  // Create decision variables for all of the animals
  let animalDecs =
      DecisionBuilder "NumberOf" {
          for animal in animals -> Continuous (0.0, infinity)
      }
     |> Map.ofSeq


  // Create an objective expression
  let objectiveExpression = 
    [ for KeyValue(animal, decVar) in animalDecs -> 
      (revenue.[animal] * decVar)
    ] |> List.sum
    
  // Create the objective
  let objective = Objective.create "MaximizeRevenue" Maximize objectiveExpression

  // Add the constraint for the total amount of pasture available
  let pastureExpr = 
    [ for animal in animals -> 
      pasture.[animal] * animalDecs.[animal]
    ] |> List.sum

  let maxPastureConstraint = Constraint.create "MaxPasture" (pastureExpr <== totalPasture)

  // Add the constraint for the total amount of feed available
  let feedExpr = 
    [ for animal in animals -> 
      feed.[animal] * animalDecs.[animal]
    ] |> List.sum

  let maxFeedConstraint = Constraint.create "MaxFeed" (feedExpr <== totalFeed)

  // Add the constraint for the total amount of labor available
  let laborExpr = 
    [ for animal in animals -> 
      labor.[animal] * animalDecs.[animal]
    ] |> List.sum

  let maxLaborConstraint = Constraint.create "MaxLabor" (laborExpr <== totalLabor)


  // Create the model and add the constraints
  let model = 
      Model.create objective
      |> Model.addConstraint maxPastureConstraint
      |> Model.addConstraint maxFeedConstraint
      |> Model.addConstraint maxLaborConstraint


  // Create the solver settings
  let settings = {
          SolverType = SolverType.CBC
          MaxDuration = 10_000L
          WriteLPFile = None
      }

  // Solve the model and save the result
  let result = Solver.solve settings model

  printfn "--Results--"

  // Print the results of the solver
  match result with
      | Suboptimal msg -> printfn "Unable to solve. Error %s" msg
      | Optimal solution ->
          for (decision, value) in solution.DecisionResults |> Map.toSeq do
              let (DecisionName name) = decision.Name
              printfn "Decision: %s\tValue: %f" name value  

File: code/csharp-visualstudio/Solutions/03-RevisedFarmModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Google.OrTools.LinearSolver;

namespace Solutions
{
  public class RevisedFarmModel
  {
    public static void Solve()
    {
      var animals = new List<string>
      {
        "cow", "pig", "chicken"
      };

      var revenue = new Dictionary<string, double>()
      {
        { "cow", 100.0 }, {"pig", 50.0 }, {"chicken", 8.0 }
      };

      var pasture = new Dictionary<string, double>()
      {
        { "cow", 1.0 }, {"pig", 0.5 }, {"chicken", 0.0 }
      };

      var feed = new Dictionary<string, double>()
      {
        { "cow", 0.0 }, {"pig", 1.0 }, {"chicken", 0.1 }
      };

      var labor = new Dictionary<string, double>()
      {
        { "cow", 9.5 }, {"pig", 8.0 }, {"chicken", 0.0 }
      };

      var totalPasture = 1000.0;
      var totalFeed = 100.0;
      var totalLabor = 7200.0;

      // Create an instance of the Solver class
      var solver = Solver.CreateSolver("FarmModel", "GLOP_LINEAR_PROGRAMMING");

      // Create variables for each type of Anaimal
      var decisions =
        animals
        .ToDictionary(x => x, x => solver.MakeNumVar(0.0, Double.MaxValue, x));

      // Create expression for Revenue
      var objExpr = 
        animals.Sum(x => decisions[x] * revenue[x]);

      // Add the expression to the solver
      solver.Maximize(objExpr);

      // Create constraint for Pasture
      var maxPasture = animals.Sum(x => decisions[x] * pasture[x]) <= totalPasture;
      solver.Add(maxPasture);

      // Create constraint for the Feed
      var maxFeed = animals.Sum(x => decisions[x] * feed[x]) <= totalFeed;
      solver.Add(maxFeed);

      // Create constraint for the Labor
      var maxLabor = animals.Sum(x => decisions[x] * labor[x]) <= totalLabor;
      solver.Add(maxLabor);

      // Solve the problem
      var resultStatus = solver.Solve();

      // Evaluate the solve status
      if (resultStatus != Solver.ResultStatus.OPTIMAL)
      {
        Console.WriteLine("The problem does not have an optimal solution!");
        return;
      }

      // Print the results
      Console.WriteLine("Solution:");
      Console.WriteLine("Objective value = " + solver.Objective().Value());
      foreach (var variable in solver.variables())
      {
        Console.WriteLine($"{variable.Name()} = {variable.SolutionValue()}");
      }
    }
  }
}
